Makefile

```diff
@@ -1,79 +1,130 @@
 #
-ESLINT		:= ./node_modules/.bin/eslint
-DOCUMENTATION		:= ./node_modules/.bin/documentation
-NSP		:= ./node_modules/.bin/nsp
-NODEUNIT	:= ./node_modules/.bin/nodeunit
-MOCHA		:= ./node_modules/.bin/mocha
-NODECOVER	:= ./node_modules/.bin/cover
-DOCS_BUILD	:= ./tools/docsBuild.js
 NPM		:= npm
-NODE		:= node
-PRETTIER		:= ./node_modules/.bin/prettier
+COVERALLS	:= $(NODE_BIN)/coveralls
+ESLINT		:= $(NODE_BIN)/eslint
+MOCHA		:= $(NODE_BIN)/mocha
+NODEUNIT	:= $(NODE_BIN)/nodeunit
+NYC		:= $(NODE_BIN)/nyc
+PRETTIER	:= $(NODE_BIN)/prettier
+DOCS_BUILD	:= $(TOOLS)/docsBuild.js
+
 
 #
-# Files
+# Files and globs
 #
-JS_FILES	 = '.'
-
-CLEAN_FILES	+= node_modules cscope.files
-
-include ./tools/mk/Makefile.defs
+PACKAGE_JSON	:= $(ROOT)/package.json
+GITHOOKS	:= $(wildcard $(GITHOOKS_SRC)/*)
+ALL_FILES	:= $(shell find $(ROOT) \
+			-not \( -path $(NODE_MODULES) -prune \) \
+			-not \( -path $(COVERAGE) -prune \) \
+			-not \( -path $(COVERAGE_RES) -prune \) \
+			-name '*.js' -type f)
+TEST_FILES	:= $(shell find $(TEST) -name '*.js' -type f)
 
 #
-# Repo-specific targets
+# Targets
 #
-.PHONY: all
-all: $(NODEUNIT) $(REPO_DEPS)
-	$(NPM) rebuild
 
-$(NODEUNIT): | $(NPM_EXEC)
-	$(NPM) install
+$(NODE_MODULES): $(PACKAGE_JSON) ## Install node_modules
+	@$(NPM) install
+	@touch $(NODE_MODULES)
 
-$(NODECOVER): | $(NPM_EXEC)
-	$(NPM) install
 
-.PHONY: cover
-cover: $(NODECOVER)
-	@rm -fr ./.coverage_data
-	$(NODECOVER) run $(NODEUNIT) ./test/*.js
-	$(NODECOVER) report html
+.PHONY: help
+help:
+	@perl -nle'print $& if m{^[a-zA-Z_-]+:.*?## .*$$}' $(MAKEFILE_LIST) \
+		| sort | awk 'BEGIN {FS = ":.*?## "}; \
+		{printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
 
-CLEAN_FILES += $(TAP) ./node_modules/nodeunit
 
-.PHONY: test
-test: $(NODEUNIT)
-	$(NODEUNIT) test/*.test.js
-	$(MOCHA) test/plugins/*.test.js
+.PHONY: githooks
+githooks: $(GITHOOKS) ## Symlink githooks
+	@$(foreach hook,\
+		$(GITHOOKS),\
+		ln -sf $(hook) $(GITHOOKS_DEST)/$(hook##*/);\
+	)
+
+
+.PHONY: lint
+lint: $(NODE_MODULES) $(ESLINT) $(ALL_FILES) ## Run lint checker (eslint).
+	@NO_STYLE=true $(ESLINT) $(ALL_FILES)
+
+
+.PHONY: style
+style: $(NODE_MODULES) $(ESLINT) $(ALL_FILES) ## Run lint checker (eslint).
+	@NO_LINT=true $(ESLINT) $(ALL_FILES)
+
+
+.PHONY: fix-style
+fix-style: $(NODE_MODULES) $(PRETTIER) ## Run prettier to auto fix style issues.
+	@$(PRETTIER) --write "**/*.js"
+
+
+.PHONY: security
+security: $(NODE_MODULES) ## Check for dependency vulnerabilities.
+	@$(NPM) audit
 
-.PHONY: nsp
-nsp: node_modules $(NSP)
-	@($(NSP) check) | true
 
 .PHONY: docs-build
-docs-build:
+docs-build: $(NODE_MODULES) $(DOCS_BUILD) ## Build documentation from JSDocs
 	@($(NODE) $(DOCS_BUILD))
 
+
 .PHONY: benchmark
-benchmark:
+benchmark: $(NODE_MODULES)
 	@(cd ./benchmark && $(NPM) i && $(NODE) index.js)
 
-include ./tools/mk/Makefile.deps
-include ./tools/mk/Makefile.targ
+
+.PHONY: prepush
+prepush: $(NODE_MODULES) lint style test ## Git pre-push hook task. Run before committing and pushing.
```

Comment:
Should this be `test` or `coverage`?

Reply:
I have it as test, since the coverage target is specified as a post build hook in .travis.yml. We could run it locally too though - don't feel strongly about it.